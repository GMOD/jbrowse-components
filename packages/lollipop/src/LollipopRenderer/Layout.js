import { readConfObject } from '@gmod/jbrowse-core/configuration'
import { doesIntersect2 } from '@gmod/jbrowse-core/util/range'

export class FloatingLayout {
  constructor({ width }) {
    if (!width) throw new Error('width required to make a new FloatingLayout')
    this.width = width
  }

  items = []

  layout = new Map()

  layoutDirty = false

  add(uniqueId, anchorLocation, width, height, data) {
    this.items.push({ uniqueId, anchorLocation, width, height, data })
    this.layoutDirty = true
  }

  /**
   * @returns {Map} of uniqueId => {x,y,anchorLocation,width,height,data}
   */
  getLayout(configuration) {
    if (!this.layoutDirty) return this.layout
    if (!configuration) throw new Error('configuration object required')

    const minY = readConfObject(configuration, 'minStickLength')

    // sort them by score ascending, so higher scores will always end up
    // stacked last (toward the bottom)
    const sorted = this.items.sort((a, b) => a.data.score - b.data.score)

    // bump them
    let maxBottom = 0
    const layoutEntries = Array(sorted.length)
    for (let i = 0; i < sorted.length; i += 1) {
      const currentItem = sorted[i]
      const { anchorLocation, width, height } = currentItem
      const start = anchorLocation - width / 2
      const end = start + width
      let top = minY
      let bottom = top + height

      // figure out how far down to put it
      for (let j = 0; j < i; j += 1) {
        const [, previouslyLayedOutItem] = layoutEntries[j]
        const {
          x: prevStart,
          y: prevTop,
          width: prevWidth,
          height: prevHeight,
        } = previouslyLayedOutItem
        const prevEnd = prevStart + prevWidth
        const prevBottom = prevTop + prevHeight
        if (
          doesIntersect2(prevStart, prevEnd, start, end) &&
          doesIntersect2(prevTop, prevBottom, top, bottom)
        ) {
          // bump this one to the bottom of the previous one
          top = prevBottom
          bottom = top + height
          j = -1 // we need to check all of them again after bumping
        }
      }

      // record the entry and update the maxBottom
      layoutEntries[i] = [
        currentItem.uniqueId,
        { ...currentItem, x: start, y: top },
      ]
      if (bottom > maxBottom) maxBottom = bottom
    }

    // try to tile them left to right all at the same level
    // if they don't fit, try to alternate them on 2 levels, then 3
    this.totalHeight = maxBottom
    this.layout = new Map(layoutEntries)
    this.layoutDirty = false
    return this.layout
  }

  getTotalHeight() {
    if (this.layoutDirty) this.getLayout()
    return this.totalHeight
  }

  toJSON() {
    return { pairs: [...this.getLayout()], totalHeight: this.getTotalHeight() }
  }

  static fromJSON() {
    throw new Error('not supported')
  }
}

export class PrecomputedFloatingLayout {
  constructor({ pairs, totalHeight }) {
    this.layout = new Map(pairs)
    this.totalHeight = totalHeight
  }

  add(uniqueId) {
    if (!this.layout.has(uniqueId))
      throw new Error(`layout error, precomputed layout is missing ${uniqueId}`)
  }

  getLayout() {
    return this.layout
  }

  getTotalHeight() {
    return this.totalHeight
  }

  static fromJSON(json) {
    return new PrecomputedFloatingLayout(json)
  }
}
