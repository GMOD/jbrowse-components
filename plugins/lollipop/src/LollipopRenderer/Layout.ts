import { readConfObject } from '@jbrowse/core/configuration'
import { doesIntersect2 } from '@jbrowse/core/util/range'
import { AnyConfigurationModel } from '@jbrowse/core/configuration/configurationSchema'

interface LayoutItem {
  uniqueId: string
  anchorLocation: number
  width: number
  height: number
  data: { score: number }
}

type LayoutEntry = LayoutItem & { x: number; y: number }

type LayoutMap = Map<string, LayoutEntry>

export class FloatingLayout {
  width: number

  totalHeight = 0

  constructor({ width }: { width: number }) {
    if (!width) throw new Error('width required to make a new FloatingLayout')
    this.width = width
  }

  items: LayoutItem[] = []

  layout: LayoutMap = new Map()

  layoutDirty = false

  add(
    uniqueId: string,
    anchorLocation: number,
    width: number,
    height: number,
    data: { score: number },
  ) {
    this.items.push({ uniqueId, anchorLocation, width, height, data })
    this.layoutDirty = true
  }

  /**
   * @returns Map of `uniqueId => {x,y,anchorLocation,width,height,data}`
   */
  getLayout(configuration?: AnyConfigurationModel) {
    if (!this.layoutDirty) return this.layout
    if (!configuration) throw new Error('configuration object required')

    const minY = readConfObject(configuration, 'minStickLength')

    // sort them by score ascending, so higher scores will always end up
    // stacked last (toward the bottom)
    const sorted = this.items.sort((a, b) => a.data.score - b.data.score)

    // bump them
    let maxBottom = 0
    const layoutEntries: [string, LayoutEntry][] = Array(sorted.length)
    for (let i = 0; i < sorted.length; i += 1) {
      const currentItem = sorted[i]
      const { anchorLocation, width, height } = currentItem
      const start = anchorLocation - width / 2
      const end = start + width
      let top = minY
      let bottom = top + height

      // figure out how far down to put it
      for (let j = 0; j < i; j += 1) {
        const [, previouslyLayedOutItem] = layoutEntries[j]
        const {
          x: prevStart,
          y: prevTop,
          width: prevWidth,
          height: prevHeight,
        } = previouslyLayedOutItem
        const prevEnd = prevStart + prevWidth
        const prevBottom = prevTop + prevHeight
        if (
          doesIntersect2(prevStart, prevEnd, start, end) &&
          doesIntersect2(prevTop, prevBottom, top, bottom)
        ) {
          // bump this one to the bottom of the previous one
          top = prevBottom
          bottom = top + height
          j = -1 // we need to check all of them again after bumping
        }
      }

      // record the entry and update the maxBottom
      layoutEntries[i] = [
        currentItem.uniqueId,
        { ...currentItem, x: start, y: top },
      ]
      if (bottom > maxBottom) maxBottom = bottom
    }

    // try to tile them left to right all at the same level
    // if they don't fit, try to alternate them on 2 levels, then 3
    this.totalHeight = maxBottom
    this.layout = new Map(layoutEntries)
    this.layoutDirty = false
    return this.layout
  }

  getTotalHeight() {
    if (this.layoutDirty)
      throw new Error('getTotalHeight does not work when the layout is dirty.')
    return this.totalHeight
  }

  serializeRegion() {
    return this.toJSON()
  }

  toJSON() {
    if (this.layoutDirty)
      throw new Error('toJSON does not work when the layout is dirty.')
    return { pairs: [...this.getLayout()], totalHeight: this.getTotalHeight() }
  }

  static fromJSON() {
    throw new Error('not supported')
  }
}

export class PrecomputedFloatingLayout {
  layout: LayoutMap

  totalHeight: number

  constructor({
    pairs,
    totalHeight,
  }: {
    pairs: [string, LayoutEntry][]
    totalHeight: number
  }) {
    this.layout = new Map(pairs)
    this.totalHeight = totalHeight
  }

  add(uniqueId: string) {
    if (!this.layout.has(uniqueId))
      throw new Error(`layout error, precomputed layout is missing ${uniqueId}`)
  }

  getLayout() {
    return this.layout
  }

  getTotalHeight() {
    return this.totalHeight
  }

  static fromJSON(
    json: ConstructorParameters<typeof PrecomputedFloatingLayout>[0],
  ) {
    return new PrecomputedFloatingLayout(json)
  }
}
