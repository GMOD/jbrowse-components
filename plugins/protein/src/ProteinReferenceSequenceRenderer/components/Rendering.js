import { readConfObject } from '@jbrowse/core/configuration'
import { PropTypes as CommonPropTypes } from '@jbrowse/core/util/types/mst'
import { objectFromEntries } from '@jbrowse/core/util'
import { contrastingTextColor } from '@jbrowse/core/util/color'
import SimpleFeature from '@jbrowse/core/util/simpleFeature'
import { observer } from 'mobx-react'
import ReactPropTypes from 'prop-types'
import React from 'react'
import aminoAcids from '../aminoAcids'
import nucleotides from '../nucleotides'
import { featuresConsensusSequence } from '../util'

function Sequence({
  getColorForLetter,
  getTitleForLetter,
  region,
  height,
  bpPerPx,
  sequence,
  lettersPerBp,
  y,
}) {
  let s = sequence.split('')
  if (region.reversed) s = s.reverse()
  const letterWidth = 1 / bpPerPx / lettersPerBp
  return (
    <g transform={`translate(0 ${y})`}>
      {s.map((letter, iter) => {
        const left = iter * letterWidth
        const fill = getColorForLetter(letter)
        return (
          <g
            key={`${region.start}-${iter}`}
            transform={`translate(${left}, 0)`}
          >
            <title>{getTitleForLetter(letter)}</title>
            <rect x={0} y={0} width={letterWidth} height={height} fill={fill} />
            {letterWidth < height * 0.8 ? null : (
              <text
                textAnchor="middle"
                x={letterWidth / 2}
                y={height * 0.8}
                fill={contrastingTextColor(fill)}
              >
                {letter}
              </text>
            )}
          </g>
        )
      })}
    </g>
  )
}

Sequence.propTypes = {
  region: CommonPropTypes.Region.isRequired,
  bpPerPx: ReactPropTypes.number.isRequired,
  y: ReactPropTypes.number,
  getTitleForLetter: ReactPropTypes.func,
  getColorForLetter: ReactPropTypes.func,
  sequence: ReactPropTypes.string.isRequired,
  height: ReactPropTypes.number.isRequired,
  lettersPerBp: ReactPropTypes.number,
}
Sequence.defaultProps = {
  y: 0,
  getTitleForLetter: () => undefined,
  getColorForLetter: () => '#ffffff',
  lettersPerBp: 1,
}

/**
 * given an iterable of features, yield only those features with
 * the requested type
 *
 * @param {string} type
 * @param {Iterable[SimpleFeature]} features
 * @returns {Iterable} of features in the iterator that have the requested type
 */
function* getFeaturesOfType(type, features) {
  for (const feature of features) {
    if (feature.get('type') === type) yield feature
  }
}

function* scaleFeatureCoordinates(scaleFactor, features) {
  for (const feature of features) {
    const json = feature.toJSON()
    json.start *= scaleFactor
    json.end *= scaleFactor
    yield new SimpleFeature(json)
  }
}
function scaleRegion(factor, region) {
  const scaled = { ...region }
  scaled.start *= factor
  scaled.end *= factor
  return scaled
}

const aaColors = objectFromEntries(
  aminoAcids.map(aaRecord => [aaRecord.letter, aaRecord.color]),
)
const aaNames = objectFromEntries(
  aminoAcids.map(aaRecord => [aaRecord.letter, aaRecord.fullName]),
)
const bpColors = objectFromEntries(
  nucleotides.map(nRecord => [nRecord.letter, nRecord.color]),
)

function Rendering(props) {
  const { bpPerPx, config, regions, features } = props
  const [region] = regions

  if (bpPerPx > 1) {
    return (
      <div style={{ whiteSpace: 'normal', padding: '2px' }}>
        Zoom in to see sequence
      </div>
    )
  }

  const height = readConfObject(config, 'height')
  const widthPx = (region.end - region.start) / bpPerPx
  const proteinSequence = featuresConsensusSequence(
    region,
    getFeaturesOfType('protein', features.values()),
  )

  const showDnaSequence = true
  const dnaSequence = featuresConsensusSequence(
    scaleRegion(3, region),
    scaleFeatureCoordinates(3, getFeaturesOfType('dna', features.values())),
  )

  return (
    <svg
      height={height + (showDnaSequence ? height : 0)}
      width={widthPx}
      style={{ fontSize: height * 0.9 }}
    >
      {showDnaSequence ? (
        <Sequence
          {...props}
          region={region}
          lettersPerBp={3}
          height={height}
          sequence={dnaSequence}
          getColorForLetter={letter => bpColors[letter] || '#ffffff'}
        />
      ) : null}
      <Sequence
        {...props}
        region={region}
        y={showDnaSequence ? height : 0}
        height={height}
        sequence={proteinSequence}
        getColorForLetter={letter => aaColors[letter] || '#ffffff'}
        getTitleForLetter={letter => aaNames[letter]}
      />
    </svg>
  )
}
Rendering.propTypes = {
  config: CommonPropTypes.ConfigSchema.isRequired,
  bpPerPx: ReactPropTypes.number.isRequired,
  regions: ReactPropTypes.arrayOf(
    ReactPropTypes.shape({
      end: ReactPropTypes.number.isRequired,
      start: ReactPropTypes.number.isRequired,
    }),
  ).isRequired,
  features: ReactPropTypes.instanceOf(Map).isRequired,
}

export default observer(Rendering)
