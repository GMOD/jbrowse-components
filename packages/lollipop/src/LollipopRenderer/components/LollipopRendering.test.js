import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import React from 'react'
import TestRenderer from 'react-test-renderer'
import ConfigSchema from '../configSchema'
import { FloatingLayout, PrecomputedFloatingLayout } from '../Layout'
import Rendering from './LollipopRendering'

// these tests do very little, let's try to expand them at some point
test('no features', () => {
  const renderer = TestRenderer.create(
    <Rendering
      width={500}
      height={500}
      region={{ refName: 'zonk', start: 0, end: 300 }}
      layout={new PrecomputedFloatingLayout({ pairs: [], totalHeight: 20 })}
      config={{}}
      bpPerPx={3}
    />,
  )
  const result = renderer.toJSON()

  expect(result).toMatchSnapshot()
})

test('one feature', () => {
  const renderer = TestRenderer.create(
    <Rendering
      width={500}
      height={500}
      region={{
        refName: 'zonk',
        start: 0,
        end: 1000,
      }}
      layout={new FloatingLayout({ width: 100 })}
      features={
        new Map([
          [
            'one',
            new SimpleFeature({ id: 'one', score: 10, start: 1, end: 3 }),
          ],
        ])
      }
      config={ConfigSchema.create({})}
      bpPerPx={3}
    />,
  )
  const result = renderer.toJSON()

  expect(result).toMatchSnapshot()
})
