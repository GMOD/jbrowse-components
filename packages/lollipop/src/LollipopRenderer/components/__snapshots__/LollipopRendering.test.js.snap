// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no features 1`] = `
<svg
  className="LollipopRendering"
  height={20}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
  width={100}
/>
`;

exports[`one feature 1`] = `
<svg
  className="LollipopRendering"
  height={16.286652959662007}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseOut={[Function]}
  onMouseOver={[Function]}
  onMouseUp={[Function]}
  style={
    Object {
      "position": "relative",
    }
  }
  width={333.3333333333333}
>
  <line
    stroke="black"
    strokeWidth={2}
    x1={0.7}
    x2={0.7}
    y1={0}
    y2={16.286652959662007}
  />
  <g>
    <title />
    <circle
      cx={0.7}
      cy={10.643326479831003}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onMouseDown={[Function]}
      onMouseEnter={[Function]}
      onMouseLeave={[Function]}
      onMouseMove={[Function]}
      onMouseOut={[Function]}
      onMouseOver={[Function]}
      onMouseUp={[Function]}
      r={5.643326479831003}
      style={
        Object {
          "fill": "green",
        }
      }
    />
  </g>
</svg>
`;
